File: py-sidecar/runtime/req_res.py
from __future__ import annotations

import asyncio
import json
from collections.abc import Awaitable, Callable, Coroutine
from typing import TYPE_CHECKING, Any

from bus_message import BusMessage
from error_log import error_log
from registry import Registry

if TYPE_CHECKING:
    from facade import Facade


class RequestStates:
    def __init__(self) -> None:
        self._pending: dict[str, asyncio.Future[Any]] = {}
        self._counter = 0

    def insert(self) -> tuple[str, asyncio.Future[Any]]:
        loop = asyncio.get_running_loop()
        future: asyncio.Future[Any] = loop.create_future()
        request_id = str(self._counter)
        self._counter += 1
        self._pending[request_id] = future
        return request_id, future

    def respond(self, request_id: str, value: Any) -> None:
        future = self._pending.pop(request_id, None)
        if future is not None and not future.done():
            future.set_result(value)

    def fail(self, request_id: str, error: str) -> None:
        future = self._pending.pop(request_id, None)
        if future is not None and not future.done():
            future.set_exception(RuntimeError(error))


PENDING_REQUESTS = RequestStates()


async def _run_handler(handler: Callable, facade: Facade, payload: Any) -> Any:
    if asyncio.iscoroutinefunction(handler):
        return await handler(facade, payload)
    return handler(facade, payload)


def get_message_handler(
    registry: Registry,
    facade: Facade,
    loop: asyncio.AbstractEventLoop,
    track: Callable[[asyncio.Task[None]], None] | None = None,
) -> Callable[[str], None]:
    event_handlers, request_handlers = registry.clone_sidecar_handlers()

    def spawn(coro: Coroutine[Any, Any, Any]) -> None:
        task = loop.create_task(coro)
        if track is not None:
            track(task)

    def on_message(line: str) -> None:
        line = line.strip()
        if not line:
            return

        try:
            msg = BusMessage.from_json(line)
        except (json.JSONDecodeError, KeyError) as exc:
            error_log(f"[PY] invalid json: {line!r} ({exc})")
            return

        if msg.kind == "event":
            handlers = event_handlers.get(msg.route, [])
            if not handlers:
                error_log(f"[PY] no handler for event: {msg.route}")
                return
            for handler in handlers:
                spawn(_run_handler(handler, facade, msg.payload))
            return

        if msg.kind == "request":
            handler = request_handlers.get(msg.route)
            if handler is None:
                error_log(f"[PY] no handler for request: {msg.route}")
                return

            async def handle_request() -> None:
                try:
                    result = await _run_handler(handler, facade, msg.payload)
                    facade.send(msg.to_response(result))
                except Exception as exc:
                    error_log(f"[PY] request handler error: {exc}")
                    facade.send(msg.to_response({"error": str(exc)}))

            spawn(handle_request())
            return

        if msg.kind == "response":
            if not msg.id:
                error_log("[PY] response message has no id")
                return
            PENDING_REQUESTS.respond(msg.id, msg.payload)
            return

        error_log(f"[PY] unknown message kind: {msg.kind}")

    return on_message

File: src-tauri/src/core/req_res.rs
use std::{
    collections::HashMap,
    future::Future,
    sync::{LazyLock, Mutex},
    time::Duration,
};

use serde_json::Value;
use tokio::sync::oneshot;
use tokio::time::timeout;

use crate::{core::Facade, error_log, sidecar::BusMessage};

use super::registry::Registry;

pub struct RequestStates {
    pending_requests: HashMap<String, oneshot::Sender<Value>>,
    counter: std::sync::atomic::AtomicUsize,
}

impl RequestStates {
    fn new() -> Self {
        Self {
            pending_requests: HashMap::new(),
            counter: std::sync::atomic::AtomicUsize::new(0),
        }
    }

    fn insert(
        &mut self,
        timeout_duration: Duration,
    ) -> (
        String,
        impl std::future::Future<Output = Result<Value, String>>,
    ) {
        let (tx, rx) = oneshot::channel::<Value>();
        let id = self
            .counter
            .fetch_add(1, std::sync::atomic::Ordering::Relaxed);
        let id = id.to_string();
        let id_clone = id.clone();
        self.pending_requests.insert(id.clone(), tx);
        let waiter = async move {
            match timeout(timeout_duration, rx).await {
                Ok(Ok(value)) => Ok(value),
                Ok(Err(_)) => Err(String::from(format!("#{} channel closed", id_clone))),
                Err(_) => Err(String::from(format!("#{} timeout", id_clone))),
            }
        };
        (id, waiter)
    }

    fn respond(&mut self, id: String, value: Value) {
        if let Some(tx) = self.pending_requests.remove(&id) {
            tx.send(value).unwrap();
        }
    }
}

static PENDING_REQUESTS: LazyLock<Mutex<RequestStates>> =
    LazyLock::new(|| Mutex::new(RequestStates::new()));

pub fn new_request(timeout_duration: Duration) -> (String, impl Future<Output = Result<Value, String>>) {
    PENDING_REQUESTS.lock().unwrap().insert(timeout_duration)
}

pub fn get_message_handler(
    registry: &Registry,
    facade: Facade,
    rt: tokio::runtime::Handle,
) -> Box<dyn Fn(String) + Send + Sync + 'static> {
    let (event_handlers, request_handlers) = registry.clone_sidecar_handlers();
    let rt_for_message = rt.clone();

    let on_message = move |message: String| {
        let message = message.trim();
        let Ok(msg) = serde_json::from_str::<BusMessage>(&message) else {
            error_log!("invalid json: {message}");
            return;
        };

        match msg.kind.as_str() {
            "event" => {
                if let Some(handlers) = event_handlers.get(&msg.route) {
                    for handler in handlers.clone() {
                        let payload = msg.payload.clone();
                        let facade = facade.clone();
                        rt_for_message.spawn(async move {
                            handler.call(facade, payload).await;
                        });
                    }
                } else {
                    error_log!("no handler for event: {}", msg.route);
                }
            }
            "request" => {
                if let Some(handler) = request_handlers.get(&msg.route).cloned() {
                    let payload = msg.payload.clone();
                    let facade = facade.clone();
                    rt_for_message.spawn(async move {
                        match handler.call(facade.clone(), payload).await {
                            Ok(res) => {
                                facade.send(msg.to_response(res));
                            }
                            Err(e) => {
                                error_log!("request handler error: {e}");
                            }
                        }
                    });
                } else {
                    error_log!("no handler for request: {}", msg.route);
                }
            }
            "response" => {
                if msg.id == "" {
                    error_log!("response message has no id");
                    return;
                }

                PENDING_REQUESTS
                    .lock()
                    .unwrap()
                    .respond(msg.id, msg.payload);
            }
            _ => error_log!("unknown message kind: {}", msg.kind),
        }
    };
    Box::new(on_message)
}

